src/main/scala/mesosphere/marathon/api/v2/AppsResource.scala

```diff
@@ -300,22 +301,16 @@ class AppsResource @Inject() (
   private def maybePostEvent(req: HttpServletRequest, app: AppDefinition) =
     eventBus.publish(ApiPostEvent(req.getRemoteAddr, req.getRequestURI, app))
 
-  private def search(cmd: String, id: String): Iterable[AppDefinition] = {
-    /* Returns true iff `a` is a prefix of `b`, case-insensitively */
-    def isPrefix(a: String, b: String): Boolean =
-      b.toLowerCase contains a.toLowerCase
+  private def search(cmd: String, id: String, label: String): Iterable[AppDefinition] = {
+    /* Returns true if `a` is a prefix of `b`, case-insensitively */
+    def isPrefix(a: String, b: String): Boolean = b.toLowerCase contains a.toLowerCase
+    val selectors = if (label != null) Some(new LabelSelectorParsers().parsed(label)) else None
 
     service.listApps().filter { app =>
-      val appMatchesCmd =
-        cmd != null &&
-          cmd.nonEmpty &&
-          app.cmd.exists(isPrefix(cmd, _))
-
-      val appMatchesId =
-        id != null &&
-          id.nonEmpty && isPrefix(id, app.id.toString)
-
-      appMatchesCmd || appMatchesId
+      val appMatchesCmd = cmd == null || (cmd != null && cmd.nonEmpty && app.cmd.exists(isPrefix(cmd, _)))
+      val appMatchesId = id == null || (id != null && id.nonEmpty && isPrefix(id, app.id.toString))
+      val appMatchesLabel = label == null || selectors.exists(_.matches(app))
+      appMatchesCmd && appMatchesId && appMatchesLabel
```

Comment:
I think that this logic is untested. If you moved `search` into a different class/object (maybe slightly refactor it), it would be quite easy to test.